### PR TITLE
feat: add coincurve instead of secp256k1

### PR DESCRIPTION
**Pull request by dni**
- changed secp256k1 to coincurve
- add testcase for verifying pubkey https://github.com/lightning/bolts/issues/1184
- added keep_payee flag for encoding
- added tests for signature model
- added test for invalid signature

### CODE REVIEW

**Comment by prusnak** [comment]
Can you also try to remove ecdsa from dependencies? Should be possible

File: tests/test_bolt11_examples.py [hunk 1]
```diff
@@ -1125,3 +1123,35 @@ def test_example_14(self):
         )
         encoded = encode(invoice, ex["private_key"])
         assert encoded == ex["payment_request"]
+
+    def test_example_15(self):
+        """
+        Verify payee signature
+        """
+        ex = {
```
Comment (motorina0) [inline tests/test_bolt11_examples.py:1131]: nitpick: json data should be extracted in a `fixtures.json` file to keep the code claner

File: tests/test_bolt11_examples.py [hunk 2]
```diff
@@ -1125,3 +1123,35 @@ def test_example_14(self):
...
+            ),
+        }
+
+        decoded = decode(ex["payment_request"])
+        assert decoded.payment_secret == ex["payment_secret"]
+        assert decoded.description == ex["description"]
+        assert decoded.payee == ex["payee"]
+        assert decoded.signature
+
+        re_encoded = encode(decoded, keep_payee=True)
```
Comment (motorina0) [inline tests/test_bolt11_examples.py:1156]: test with `keep_payee=False` too

File: tests/test_bolt11_examples.py [hunk 3]
```diff
@@ -1125,3 +1123,35 @@ def test_example_14(self):
...
+            "private_key": (
+                "e126f68f7eafcc8b74f54d269fe206be715000f94dac067d1c04a8ca3b2db734"
+            ),
+        }
+
+        decoded = decode(ex["payment_request"])
+        assert decoded.payment_secret == ex["payment_secret"]
+        assert decoded.description == ex["description"]
+        assert decoded.payee == ex["payee"]
+        assert decoded.signature
```
Comment (motorina0) [inline tests/test_bolt11_examples.py:1154]: can the `pubkey` be extracted from the `signature` and `asserted`?

File: tests/test_signature.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    "public_key": "031a7a2db2e0c0fbfcfe62e23b6262b9f2b76bc01b022d17b2f9b49c7bc22310fd",
+    "signature": (
+        "202bea309bab7aeff223ad2890d42ca00bc5673c88cab47420387100f8939fb26a"
+        "dea33c4e1748b3a56f87b3f9b1b001ffdc82169809782a02bd03ae2d1cbdc700"
+    ),
+}
+
+
+class TestBolt11Signature:
+    def test_recovers_public_key(self):
```
Comment (motorina0) [inline tests/test_signature.py:16]: why are there 3 tests? can't all assertions be made on the first one?

**Review by motorina0 — APPROVED** [review]
LGTM

see comments

File: tests/test_bolt11_examples.py (hunk 2 above)
Comment (dni) [reply tests/test_bolt11_examples.py:1156]: false is default and used in all the other examples

**Comment by prusnak** [comment]
> Can you also try to remove ecdsa from dependencies and rewrite its usage with coincurve? Should be possible

Hm, it seems that coincurve does not provide an easy way how to use a signature in hex-string format (not DER).

We could use the following to convert hex-string sig to DER, but it does not help us with getting rid of the dependency :-/

``` python
r, s = sigdecode_string(signature_data, SECP256k1.order)
der_sig = sigencode_der_canonize(r, s, SECP256k1.order)
```

**Comment by dni** [comment]
> > Can you also try to remove ecdsa from dependencies and rewrite its usage with coincurve? Should be possible
> 
> Hm, it seems that coincurve does not provide an easy way how to use a signature in hex-string format (not DER).
> 
> We could use the following to convert hex-string sig to DER, but it does not help us with getting rid of the dependency :-/
> 
> ```python
> from ecdsa import SECP256k1
> from ecdsa.util import sigdecode_string, sigencode_der_canonize
> 
> r, s = sigdecode_string(signature_data, SECP256k1.order)
> der_sig = sigencode_der_canonize(r, s, SECP256k1.order)
> ```

yep, took me a bit to figure it out.

```
        sig = deserialize_recoverable(self.signature_data)
        sig = recoverable_convert(sig)
        sig = cdata_to_der(sig)
        if not verify_signature(
            sig, message(self.hrp, self.signing_data), bytes.fromhex(payee)
        ):
            raise ValueError("Invalid signature")

```

**Pull request merged**